Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp

```diff
@@ -0,0 +1,825 @@
+        && elementType.type.as<Wasm::PackedType>() == Wasm::PackedType::I16;
+}
+
+#define PASTE(x, y) x##y
+#define CONCAT(x, y) PASTE(x, y)
+
+// The naming scheme. Changes should be matched in InPlaceInterpreter.asm
+// macros wasmBuiltinCallTrampoline and defineWasmBuiltinCallTrampoline.
+
+#define BUILTIN_TRAMPOLINE(setName, builtinName) CONCAT(wasm_builtin_trampoline__, BUILTIN_FULL_NAME(setName, builtinName))
+#define BUILTIN_WASM_ENTRY(setName, builtinName) CONCAT(wasm_builtin__, BUILTIN_FULL_NAME(setName, builtinName))
+#define BUILTIN_JS_ENTRY(setName, builtinName) CONCAT(wasm_builtinJS__, BUILTIN_FULL_NAME(setName, builtinName))
+#define BUILTIN_IMPLEMENTATION(setName, builtinName) CONCAT(wasm_builtin_impl__, BUILTIN_FULL_NAME(setName, builtinName))
+
+#define DEFINE_BUILTIN_IMPLEMENTATION(setName, builtinName, ...) \
+    static ALWAYS_INLINE EncodedJSValue BUILTIN_IMPLEMENTATION(setName, builtinName)(__VA_ARGS__)
+
+#define DEFINE_BUILTIN_IMPLEMENTATION_I32(setName, builtinName, ...) \
+    static ALWAYS_INLINE int32_t BUILTIN_IMPLEMENTATION(setName, builtinName)(__VA_ARGS__)
+
+#define DECLARE_BUILTIN_TRAMPOLINE(setName, builtinName) \
+    extern "C" EncodedJSValue BUILTIN_TRAMPOLINE(setName, builtinName)()
+
+// A host wasm function that returns an externref (EncodedJSValue)
+#define DEFINE_WASM_BUILTIN_HOST_FUNCTION(setName, builtinName, ...) \
+    extern "C" EncodedJSValue SYSV_ABI BUILTIN_WASM_ENTRY(setName, builtinName)(__VA_ARGS__) REFERENCED_FROM_ASM WTF_INTERNAL; \
+    EncodedJSValue SYSV_ABI BUILTIN_WASM_ENTRY(setName, builtinName)(__VA_ARGS__)
+
+#define THROW_EXCEPTION \
+    do { \
+        JSObject* error = createJSWebAssemblyRuntimeError(globalObject, globalObject->vm(), Wasm::ExceptionType::IllegalArgument); \
+        throwException(globalObject, scope, error); \
+        return encodedJSValue(); \
+    } while (false)
+
+#define IF_EXCEPTION_RETHROW_AND_RETURN(scope) \
+    if (scope.exception()) \
+        THROW_EXCEPTION;
+
+// For use in JS wrappers, assuming 'valueVar' is defined at the expansion site.
+// Expand into code defining 'i32Var' and setting it to valueVar's value converted to i32.
+// The function using it should establish a throw scope.
+#define JSVALUE_TO_I32(valueVar, i32Var) \
+    int32_t i32Var; \
+    if (valueVar.isInt32()) [[likely]] { \
+        i32Var = valueVar.asInt32(); \
+    } else { \
+        double doubleVar = valueVar.toIntegerOrInfinity(globalObject); \
+        IF_EXCEPTION_RETHROW_AND_RETURN(scope); \
+        i32Var = static_cast<int32_t>(doubleVar); \
+    }
+
+// For use in JS wrappers, assuming 'var' is defined at the expansion site and contains a JSValue.
+// Expands into code that verifies that the variable contains a (non-null) array of mutable i16s.
+#define ENSURE_ARRAY_MUT_I16(var) \
+    { \
+        auto* tmpArray = static_cast<JSWebAssemblyArray*>((var).getObject()); \
+        if (!tmpArray || !isMutI16Array(tmpArray)) \
+            THROW_EXCEPTION; \
+    }
```

Comment:
This seems very wrong since the object could be any object not a `JSWebAssemblyArray*`. So the downcast will covert a potentially attacker controlled JSObject into a `JSWebAssemblyArray` then load some bytes from its structure for the `isMutI16Array` check.

This should be something like:

```
auto* tmpArray = jsDynamicCast<JSWebAssemblyArray*>(jsValue);
if (!tmpArray || !isMutI16Array(tmpArray))
    THROW_EXCEPTION;    
```